File: backend/src/main/java/com/home/backend/service/CostCalculator.java
package com.home.backend.service;

import com.home.backend.domain.AnalysisLog;
import com.home.backend.domain.MenuType;
import com.home.backend.domain.Menu;
import com.home.backend.dto.AnalysisDto;
import com.home.backend.dto.AnalysisDto.BreakdownItem;
import com.home.backend.repository.AnalysisLogRepository;
import com.home.backend.repository.MenuRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Slf4j
@Service
@RequiredArgsConstructor
public class CostCalculator {

    @Value("${home.min-wage}")
    private int minWage;

    private final MenuRepository menuRepository;
    private final AnalysisLogRepository analysisLogRepository;
    private final ExternalApiService externalApiService; 

    @Transactional
    public AnalysisDto.Response calculate(String userId, AnalysisDto.Request req) {
        
        String tempName = req.getMenuName();
        if (tempName != null && tempName.contains(",")) {
            tempName = tempName.split(",")[0].trim();
        }
        final String targetName = tempName;   
        
        Menu menu = menuRepository.findByMenuName(targetName)
                .orElseGet(() -> createDynamicFallbackMenu(targetName, null, req.getFoodPrice()));
      
        int lack = Math.max(0, req.getMinOrder() - req.getFoodPrice());
        int deliveryCost = req.getFoodPrice() + lack + req.getDeliveryFee();

        Integer mealkitCost = null;
        int mkPrice = 0, mkOppCost = 0;
        
        if (menu.getMenuType() != MenuType.DELIVERY_ONLY) {
            mkPrice = resolveMealKitPrice(targetName, req, menu);
            int kitMin = req.getKitMin() != null ? req.getKitMin() : 15;
            mkOppCost = (int) Math.round((kitMin / 60.0) * 0.2 * minWage);
            mealkitCost = mkPrice + mkOppCost;
        }

        Integer cookingCost = null;
        int ingCost = 0, ckOppCost = 0, ckLaborCost = 0, ckToolCost = 0;
        
        if (menu.getMenuType() == MenuType.ALL) {
            ingCost = resolveIngredientCost(targetName, req, menu);
            int cookMin = req.getCookMin() != null ? req.getCookMin() : menu.getDefaultCookMin();
            int laborMin = req.getLaborMin() != null ? req.getLaborMin() : 15;
            ckToolCost = req.getToolCost() != null ? req.getToolCost() : 0;

            ckOppCost = (int) Math.round((cookMin / 60.0) * 0.2 * minWage);
            ckLaborCost = (int) Math.round((laborMin / 60.0) * minWage);
            cookingCost = ingCost + ckOppCost + ckLaborCost + ckToolCost;
        }
        Map<String, List<BreakdownItem>> bd = new LinkedHashMap<>();
        
        bd.put("delivery", List.of(
                new BreakdownItem("기본가 및 옵션가", req.getFoodPrice()),
                new BreakdownItem("최소주문 미달 부족금액", lack),
                new BreakdownItem("배달팁", req.getDeliveryFee())
        ));

        if (mealkitCost != null) {
            bd.put("mealkit", List.of(
                    new BreakdownItem("밀키트 구매가", mkPrice),
                    new BreakdownItem("조리 기회비용", mkOppCost)
            ));
        }

        if (cookingCost != null) {
            bd.put("cooking", List.of(
                    new BreakdownItem("순수 식재료비", ingCost),
                    new BreakdownItem("조리 기회비용", ckOppCost),
                    new BreakdownItem("가사노동비(설거지 등)", ckLaborCost),
                    new BreakdownItem("소모 도구 비용", ckToolCost)
            ));
        }


        String bestOpt = getBestOption(deliveryCost, mealkitCost, cookingCost);
        
        AnalysisLog logData = AnalysisLog.create(
                userId, req.getMenuName(), bestOpt,
                getChosenCost(bestOpt, deliveryCost, mealkitCost, cookingCost),
                deliveryCost, mealkitCost, cookingCost,
                Math.max(0, deliveryCost - getChosenCost(bestOpt, deliveryCost, mealkitCost, cookingCost))
        );

        if (!"guest".equals(userId)) {
            analysisLogRepository.save(logData);
        }

        return AnalysisDto.Response.builder()
                .delivery(deliveryCost)
                .mealkit(mealkitCost)
                .cooking(cookingCost)
                .best(bestOpt)
                .saving(logData.getSaving())
                .breakdown(bd) 
                .build();
    }

    private int resolveIngredientCost(String menuName, AnalysisDto.Request req, Menu menu) {
        // 1순위: 사용자가 직접 입력한 값이 있으면 우선 적용
        if (req.getIngredientCost() != null && req.getIngredientCost() > 0) return req.getIngredientCost();
        
        // 2순위: 백엔드에서 외부 API(농촌진흥청/네이버 등)를 직접 호출
        try {
            Integer apiCost = externalApiService.fetchIngredientCost(menuName);
            if (apiCost != null && apiCost > 0) return apiCost;
        } catch (Exception e) {
            log.warn("외부 API 식재료비 조회 실패, DB 데이터로 폴백합니다. 메뉴: {}", menuName);
        }

        return menu.getIngredientCost() != null ? menu.getIngredientCost() : (int)(req.getFoodPrice() * 0.40);
    }

    private int resolveMealKitPrice(String menuName, AnalysisDto.Request req, Menu menu) {
        if (req.getKitPrice() != null && req.getKitPrice() > 0) return req.getKitPrice();
        
        try {
            Integer apiCost = externalApiService.fetchMealKitPrice(menuName);
            if (apiCost != null && apiCost > 0) return apiCost;
        } catch (Exception e) {
            log.warn("외부 API 밀키트비 조회 실패, DB 데이터로 폴백합니다. 메뉴: {}", menuName);
        }

        return menu.getKitPrice() != null ? menu.getKitPrice() : (int)(req.getFoodPrice() * 0.55);
    }

    private Menu createDynamicFallbackMenu(String name, String subCat, int foodPrice) {
        return Menu.builder()
                .menuName(name)
                .subCategory(subCat != null ? subCat : "기타")
                .menuType(MenuType.ALL) 
                .deliveryPrice((int)(foodPrice * 1.1))
                .minOrder(15000)
                .ingredientCost((int)(foodPrice * 0.4))
                .defaultCookMin(20)
                .build();
    }

    private String getBestOption(int delivery, Integer mealkit, Integer cooking) {
        Map<String, Integer> costs = new LinkedHashMap<>();
        costs.put("delivery", delivery);
        if (mealkit != null) costs.put("mealkit", mealkit);
        if (cooking != null) costs.put("cooking", cooking);
        return costs.entrySet().stream()
                .min(Map.Entry.comparingByValue())
                .map(Map.Entry::getKey).orElse("delivery");
    }

    private int getChosenCost(String best, int delivery, Integer mealkit, Integer cooking) {
        return switch (best) {
            case "mealkit" -> mealkit != null ? mealkit : delivery;
            case "cooking" -> cooking != null ? cooking : delivery;
            default -> delivery;
        };
    }
}

File: frontend/src/utils/costCalculator.js
import { apiCalculateCost } from "../api/apiClient"; 

// 백엔드 API  예외사항 Fallbackd으로 처리 
export async function calcCosts(menu, input, ingredientList = []) {
  try {
    const res = await apiCalculateCost({
      menuName: menu.name, 
      foodPrice: input.foodPrice, 
      deliveryFee: input.deliveryFee, 
      minOrder: input.minOrder, 
      kitPrice: input.kitPrice, 
      kitMin: input.kitMin, 
      cookMin: input.cookMin, 
      laborMin: input.laborMin, 
      toolCost: input.toolCost, 
      ingredientList, 
    });

    return res; 

  } catch (error) {
    console.error("[calcCosts] 백엔드 엔진 호출 실패, 로컬 연산으로 fallback:", error); 
    return calcCostsLocal(menu, input); 
  }
}

const MIN_WAGE = 10320; 

export function calcCostsLocal(menu, input) {
  const foodPrice = input.foodPrice || 15000;
  const deliveryFee = input.deliveryFee || 0;
  const minOrder = input.minOrder || 0;

  // API 연동 실패같은 예외사항 발생 시에 배달 가격 비례식 추정 단가 연산 (55%, 40%)
  const kitPrice = (input.kitPrice && input.kitPrice > 0) ? input.kitPrice : Math.round(foodPrice * 0.55);
  const ingredientCost = (menu.ingredientCost && menu.ingredientCost > 0) ? menu.ingredientCost : Math.round(foodPrice * 0.40);

  const kitMin = (input.kitMin && input.kitMin > 0) ? input.kitMin : 15;
  const cookMin = (input.cookMin && input.cookMin > 0) ? input.cookMin : 20;
  const laborMin = (input.laborMin && input.laborMin > 0) ? input.laborMin : 10;
  const toolCost = input.toolCost || 0;

  //최소주문에 도달하기 위해 모자라서 추가해야 하는 부족 금액
  const missingCost = Math.max(0, minOrder - foodPrice);

  // 배달 최종 비용 = 선택 메뉴 가격 + 최소주문 미달 분 부족 금액 + 배달팁
  const deliveryCost = foodPrice + missingCost + deliveryFee; 

  const mType = menu.menuType || menu.type || "ALL"; 

  const mealKitCost = mType !== "DELIVERY_ONLY" 
    ? kitPrice + (kitMin / 60) * 0.2 * MIN_WAGE 
    : null; 

  const cookingCost = mType === "ALL" 
    ? ingredientCost + (cookMin / 60) * 0.2 * MIN_WAGE
      + (laborMin / 60) * MIN_WAGE + toolCost 
    : null; 

  const available = { delivery: deliveryCost, mealkit: mealKitCost, cooking: cookingCost }; 
  const filtered = Object.entries(available).filter(([, v]) => v !== null); 
  const best = filtered.reduce((a, b) => a[1] < b[1] ? a : b)[0]; 

  const breakdown = {
    delivery: [
      { label: "선택 메뉴 기본가 및 옵션가", value: foodPrice }, 
      { label: "최소주문 미달 분 부족 금액", value: missingCost }, 
      { label: "배달팁", value: deliveryFee }, 
    ],
    mealkit: mealKitCost !== null ? [ 
      { label: "밀키트 1인분 환산가 합산", value: kitPrice }, 
      { label: "기회비용 (조리시간 × 0.2 × 최저임금)", value: Math.round((kitMin / 60) * 0.2 * MIN_WAGE) }, 
    ] : [],
    cooking: cookingCost !== null ? [ 
      { label: "식재료비", value: ingredientCost }, 
      { label: "기회비용 (조리시간 × 0.2 × 최저임금)", value: Math.round((cookMin / 60) * 0.2 * MIN_WAGE) }, 
      { label: "가사노동비 (설거지시간 × 0.2 x 최저임금)", value: Math.round((laborMin / 60) * 0.2 * MIN_WAGE) }, 
      { label: "도구 비용", value: toolCost }, 
    ] : [],
  };

  return {
    delivery: Math.round(deliveryCost), 
    mealkit: mealKitCost !== null ? Math.round(mealKitCost) : null, 
    cooking: cookingCost !== null ? Math.round(cookingCost) : null, 
    best, 
    saving: Math.max(0, deliveryCost - (available[best] || deliveryCost)), 
    breakdown, 
  };
}